frontend/src/metabase/lib/expressions/pratt/compiler.ts

```diff
@@ -0,0 +1,291 @@
+import {
+  /* ALL_ASTYPES */ ADD,
+  FIELD,
+  LOGICAL_AND,
+  CALL,
+  EQUALITY,
+  NUMBER,
+  LOGICAL_OR,
+  COMPARISON,
+  GROUP,
+  MULDIV_OP,
+  STRING,
+  SUB,
+  NEGATIVE,
+  LOGICAL_NOT,
+  IDENTIFIER,
+  ROOT,
+  ARG_LIST,
+} from "./syntax";
+import { assert, NodeType, Node, CompileError } from "./types";
+
+export type Expr = number | string | ([string, ...Expr[]] & { node?: Node });
+export type CompilerPass = (expr: Expr) => Expr;
+
+export interface Options {
+  getMBQLName(expressionName: string): string | undefined;
+  passes?: CompilerPass[];
+}
+
+type CompileFn = (node: Node, opts: Options) => Expr;
+
+export function compile(node: Node, opts: Options): Expr {
+  assert(node.type === ROOT, "Must be root node");
+  if (node.children.length > 1) {
+    throw new CompileError("Unexpected expression", {
```

Comment:
The message needs to support i18n, either here or a follow-up PR.
Check `ttag` in the `recursive-parser` for the example.

Reply:
I think that should be a separate PR that includes other stuff we forgot in custom expression like [`process.js`](https://github.com/metabase/metabase/blob/53f40966d7d994d48ea9d0bd87b3797fa453fc10/frontend/src/metabase/lib/expressions/process.js#L15)